parsl/usage_tracking/usage.py

```diff
@@ -110,28 +110,53 @@ def __init__(self, dfk, port=50077,
         self.python_version = "{}.{}.{}".format(sys.version_info.major,
                                                 sys.version_info.minor,
                                                 sys.version_info.micro)
-        self.tracking_enabled = self.check_tracking_enabled()
-        logger.debug("Tracking status: {}".format(self.tracking_enabled))
-
-    def check_tracking_enabled(self):
-        """Check if tracking is enabled.
-
-        Tracking will be enabled unless either of these is true:
+        self.tracking_level = self.check_tracking_level()
+        self.start_time = None
+        logger.debug("Tracking level: {}".format(self.tracking_level))
+
+    def check_tracking_level(self) -> int:
+        """Check if tracking is enabled and return level.
+
+        Checks the following:
+            1. PARSL_TRACKING environment variable
+                - Possible values:
+                    ["true", "false", "True", "False", "0", "1", "2", "3"]
+                - Other values are treated as Level 0 (disabled)
+
+            2. usage_tracking in Config
+                - Possible values: [True, False, 0, 1, 2, 3]
+
+            Parsl will choose the lowest level that is specified by any of the parameters, and 0 if none of them are set
+
+            True/False values are treated as Level 1/Level 0 respectively.
+
+        Returns: int
+            - 0 : Tracking is disabled
+            - 1 : Tracking is enabled with level 1
+                  Share info about Parsl version, Python version, platform
+            - 2 : Tracking is enabled with level 2
+                  Share info about config + level 1
+            - 3 : Tracking is enabled with level 3
+                  Share info about app count, app fails, execution time + level 2
+        """
+        inf = sys.maxsize
 
-            1. dfk.config.usage_tracking is set to False
-            2. Environment variable PARSL_TRACKING is set to false (case insensitive)
+        envvar_level = inf
+        envvar = str(os.environ.get("PARSL_TRACKING", 0)).lower()
+        if envvar in {"true", "false"}:
+            envvar_level = 1 if envvar == "true" else 0
 
-        """
-        track = True
+        elif envvar in {"0", "1", "2", "3"}:
+            envvar_level = int(envvar)
```

Comment:
I'm a bit uncomfortable with the style of this `if` statement:

i) try to have an `else` statement, unless there is some reason for not having one: That `if` won't fire always (that's why it is there) and if you write an explicit `else`, you can be clearer about what you intend to happen in that case. For example, in this case, I think the `else` behaviour is that we end up with envvar_level set to inf.

So you can set envvar_level to that in an else statement instead of assigning it first and then adjusting it later.

(that's a functional programming style that I've picked up from eg Haskell and Rust)

ii) In line 147, avoid doing an if than then immediately contains another if: 
the true and false cases are different, so make them different top level clauses in this if statement

So you could end up with a sequence of if/elif/elif/.../else where envvar_level is explicitly assigned in each of those clauses.

then, iii) we can look at that `else envvar_level = inf`, and ask if that is the right behaviour for every other value of PARSL_TRACKING that a user might specify?

Reply:
I've refactored the code to address suggestions from i and ii. 

For point iii, I've assigned `envvar_level = inf` when the `PARSL_TRACKING` value does not match
any conditions specified in the if statements, as we anticipate only the values specified by these conditions to be valid.

Reply:
I absolutely anticipate users setting PARSL_USAGE_TRACKING to other values, for example, `trrue` as a typo, or 99999 to get the maximum tracking by using the biggest number they can imagine is plausible.

It would be good to lead users to a solution to why this code is ignoring their specified tracking level, rather than silently failing: at the very least a `logger.warn`, but I think because this code ultimately runs in DFK startup (`DataFlowKernel.__init__` -> `UsageTracker.__init__` -> `UsageTracker.check_tracking_level`, it would be good to raise a configuration exception in that case.

Reply:
Added Exceptions for invalid values for PARSL_TRACKING set and also for invalid values set in the config